### ISSUE TITLE
Transfer Learning  - [패캠]김기현의 딥러닝을 활용한 자연어처리

### ISSUE THREAD

**Issue by sallyy1** [sallyy1/NLU-NLP#13]
- Motivations
 : 데이터가 다르더라고 같은 이미지 or 텍스트를 활용한 Task 간에는 공통된 Feature들이 존재할 것이다.

- Transfer Learning이란? 
: 어떠한 문제를 푸는 동안 얻은 knowledge들을 **“다르지만 연관된 문제”**를 푸는 데 적용하는 것
-Transfer learning is a research problem in machine learning that focuses on storing knowledge gained while solving one problem and applying it to a different but related problem.
-제로베이스(Random Initialized 하여)로 내가 모은 적은 양의 Target Dataset을 훈련시키는 것 보다 < 큰 Dataset(예: ImageNet)을 가지고 기훈련되어있는 pretrained한 네트워크의 Weight를 Load하여 나의 Task에 재학습하는 것이 **학습도 빠르며 수렴도 빠르다** 
_-NLP 예: BERT, GPT 아키텍쳐 기반 ,,_

![image](https://user-images.githubusercontent.com/70500841/106156209-99fe5b80-61c4-11eb-994c-e76cde39b3fc.png)

- 사용 방법
-1) Set seed weights and train as normal (:랜덤 시드 대신 훈련된 시드를 사용해 훈련)
: train from scratch (random init) - 아무것도 안 썼을 경우
-(추천) 2) Fix loaded weights and train unloaded parts (:불러온 Weight는 fix하고 나머지 부분(불러올 수 없었던 Train unloaded parts)들을 훈련) (*”Freeze”*)
: train with freezer pretrained weights - ‘—freeze’와 ‘—use_pretrained’ 둘다 썼을 경우
-3) Train with different learning rate on each part (:불러온 파트는 Learning Rate를 작게 하고, Unloaded 파트는 Learning Rate를 크게하여 훈련)
: train from pretrained weights - ‘—use_pretrained’ 만 썼을 경우

- Reference 문서
-https://pytorch.org/tutorials/beginner/transfer_learning_tutorial.html
-https://pytorch.org/tutorials/beginner/finetuning_torchvision_models_tutorial.html

* 실습 (bash)
```
$ python ./train.py --asdf

usage: train.py [-h] --model_fn MODEL_FN [--gpu_id GPU_ID]
                [--train_ratio TRAIN_RATIO] [--valid_ratio VALID_RATIO]
                [--test_ratio TEST_RATIO] [--batch_size BATCH_SIZE]
                [--n_epochs N_EPOCHS] [--verbose VERBOSE]
                [--model_name MODEL_NAME] [--dataset_name DATASET_NAME]
                [--n_classes N_CLASSES] [--freeze] [--use_pretrained]
train.py: error: the following arguments are required: --model_fn


$ python ./train.py --model_fn resnet.pth --gpu_id -1 --n_epochs 20 --model_name resnet --dataset_name catdog --n_classes 2 --freeze --use_pretrained


Namespace(batch_size=256, dataset_name='catdog', freeze=True, gpu_id=-1, model_fn='resnet.pth', model_name='resnet', n_classes=2, n_epochs=20, test_ratio=0.2, train_ratio=0.6, use_pretrained=True, valid_ratio=0.2, verbose=2)
Train: 15000
Valid: 5000
Test: 5000
ResNet(
  (conv1): Conv2d(3, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
  (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (relu): ReLU(inplace=True)
  (maxpool): MaxPool2d(kernel_size=3, stride=2, padding=1, dilation=1, ceil_mode=False)
  (layer1): Sequential(
    (0): BasicBlock(
      (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
    (1): BasicBlock(
      (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
    (2): BasicBlock(
      (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
  )
  (layer2): Sequential(
    (0): BasicBlock(
      (conv1): Conv2d(64, 128, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (downsample): Sequential(
        (0): Conv2d(64, 128, kernel_size=(1, 1), stride=(2, 2), bias=False)
        (1): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
    )
    (1): BasicBlock(
      (conv1): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
    (2): BasicBlock(
      (conv1): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
    (3): BasicBlock(
      (conv1): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
  )
  (layer3): Sequential(
    (0): BasicBlock(
      (conv1): Conv2d(128, 256, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (downsample): Sequential(
        (0): Conv2d(128, 256, kernel_size=(1, 1), stride=(2, 2), bias=False)
        (1): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
    )
    (1): BasicBlock(
      (conv1): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
    (2): BasicBlock(
      (conv1): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
    (3): BasicBlock(
      (conv1): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
    (4): BasicBlock(
      (conv1): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
    (5): BasicBlock(
      (conv1): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
  )
  (layer4): Sequential(
    (0): BasicBlock(
      (conv1): Conv2d(256, 512, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (downsample): Sequential(
        (0): Conv2d(256, 512, kernel_size=(1, 1), stride=(2, 2), bias=False)
        (1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
    )
    (1): BasicBlock(
      (conv1): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
    (2): BasicBlock(
      (conv1): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
  )
  (avgpool): AdaptiveAvgPool2d(output_size=(1, 1))
  (fc): Linear(in_features=512, out_features=2, bias=True)
)
Adam (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    eps: 1e-08
    lr: 0.001
    weight_decay: 0
)
NLLLoss()

  2%|█▌                                                                         Epoch [1/20]:   2%|█▎                                                           Epoch [1/20]:   2%|▎                   | 1/59 [0Epoch [1/20]:   2%|▎                   | 1/59 [01:37<?, ?it/s, accuracy=0.422, loss=-.343, |g_param|=15.Epoch [1/20]:   2%|▎                   | 1/59 [01:37<?, ?it/s, accuracy=0.421, loss=-.351, |g_param|=15.Epoch [1/20]:   3%|▎         | 2/59 [01:37<1:32:17, 97.14s/it, accuracy=0.421, loss=-.351, |g_param|=15.Epoch [1/20]:   3%|▎         | 2/59 [02:48<1:32:17, 97.14s/it, accuracy=0.421, loss=-.351, |g_param|=15.Epoch [1/20]:   3%|▎          | 2/59 [02:48<1:32:17, 97.14s/it, accuracy=0.422, loss=-.37, |g_param|=15.Epoch [1/20]:   5%|▌          | 3/59 [02:48<1:16:28, 81.93s/it, accuracy=0.422, loss=-.37, |g_param|=15.Epoch [1/20]:   5%|▌          | 3/59 [03:53<1:16:28, 81.93s/it, accuracy=0.422, loss=-.37, |g_param|=15.Epoch [1/20]:   5%|▌         | 3/59 [03:53<1:16:28, 81.93s/it, accuracy=0.421, loss=-.397, |g_param|=15.Epoch [1/20]:   7%|▋         | 4/59 [03:53<1:08:14, 74.45s/it, accuracy=0.421, loss=-.397, |g_param|=15.Epoch [1/20]:   7%|▋         | 4/59 [04:54<1:08:14, 74.45s/it, accuracy=0.421, loss=-.397, |g_param|=15.Epoch [1/20]:   7%|▋         | 4/59 [04:54<1:08:14, 74.45s/it, accuracy=0.422, loss=-.434, |g_param|=15.Epoch [1/20]:   8%|▊         | 5/59 [04:54<1:02:08, 69.04s/it, accuracy=0.422, loss=-.434, |g_param|=15.Epoch [1/20]:   8%|▊         | 5/59 [05:53<1:02:08, 69.04s/it, accuracy=0.422, loss=-.434, |g_param|=15.Epoch [1/20]:   8%|▊         | 5/59 [05:53<1:02:08, 69.04s/it, accuracy=0.422, loss=-.479, |g_param|=15.Epoch [1/20]:  10%|█▏          | 6/59 [05:53<57:49, 65.47s/it, accuracy=0.422, loss=-.479, |g_param|=15.Epoch [1/20]:  10%|█▏          | 6/59 [06:50<57:49, 65.47s/it, accuracy=0.422, loss=-.479, |g_param|=15.Epoch [1/20]:  10%|█▏          | 6/59 [06:50<57:49, 65.47s/it, accuracy=0.422, loss=-.532, 
```
